application/controllers/ServicesController.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace Icinga\Module\Graphite\Controllers;
+
+use GuzzleHttp\Psr7\ServerRequest;
+use Icinga\Module\Graphite\Util\TimeRangePickerTools;
+use Icinga\Module\Graphite\Web\Controller\TimeRangePickerTrait;
+use Icinga\Module\Graphite\Web\Widget\Graphs\Icingadb\IcingadbGraph;
+use Icinga\Module\Icingadb\Model\Service;
+use Icinga\Module\Icingadb\Web\Control\SearchBar\ObjectSuggestions;
+use Icinga\Module\Icingadb\Web\Controller;
+use ipl\Html\HtmlString;
+use ipl\Stdlib\Filter;
+use ipl\Web\Control\LimitControl;
+use ipl\Web\Control\SortControl;
+use ipl\Web\Url;
+
+class ServicesController extends Controller
+{
+    use TimeRangePickerTrait;
+
+    public function indexAction()
+    {
+        // shift graph params to avoid exception
+        foreach (['graphs_limit', 'graph_range', 'graph_start', 'graph_end'] as $param) {
+            $this->params->shift($param);
+        }
+
+        $this->setTitle(t('Services'));
+
+        $db = $this->getDb();
+
+        $services = Service::on($db)->with(['state']);
```

Comment:
Are you calling `with` just because of the filter? If so, and you don't need access to the state afterwards, don't call it. You're needlessly fetching more data otherwise.

Reply:
But... what if there's a macro like `service.state.timeout`. That would result in a query per object then. (25 additional queries per page) Not good. Maybe just forget my comment above, and even add `host` to this list, for macros like `host.name` :)